api/v1beta1/temporalschedule_types.go

```diff
@@ -0,0 +1,679 @@
+
+	// Step to be take between each value.
+	// Defaults to 1.
+	//
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=12
+	Step int32 `json:"step,omitempty"`
+}
+
+func (r ScheduleMonthRange) GetStart() int32 { return r.Start }
+func (r ScheduleMonthRange) GetEnd() int32   { return r.End }
+func (r ScheduleMonthRange) GetStep() int32  { return r.Step }
+
+// If end < start, then end is interpreted as
+// equal to start. This means you can use a Range with start set to a value, and
+// end and step unset to represent a single value.
+type ScheduleYearRange struct {
+	// Start of the range (inclusive).
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=1970
+	Start int32 `json:"start,omitempty"`
+
+	// End of the range (inclusive).
+	// Defaults to start.
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=1970
+	End int32 `json:"end,omitempty"`
+
+	// Step to be take between each value.
+	// Defaults to 1.
+	//
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	Step int32 `json:"step,omitempty"`
+}
+
+func (r ScheduleYearRange) GetStart() int32 { return r.Start }
+func (r ScheduleYearRange) GetEnd() int32   { return r.End }
+func (r ScheduleYearRange) GetStep() int32  { return r.Step }
+
+// If end < start, then end is interpreted as
+// equal to start. This means you can use a Range with start set to a value, and
+// end and step unset to represent a single value.
+type ScheduleDayOfWeekRange struct {
+	// Start of the range (inclusive).
+	// Defaults to 0.
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=6
+	Start int32 `json:"start,omitempty"`
+
+	// End of the range (inclusive).
+	// Defaults to start.
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=6
+	End int32 `json:"end,omitempty"`
+
+	// Step to be take between each value.
+	// Defaults to 1.
+	//
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=6
+	Step int32 `json:"step,omitempty"`
+}
+
+func (r ScheduleDayOfWeekRange) GetStart() int32 { return r.Start }
+func (r ScheduleDayOfWeekRange) GetEnd() int32   { return r.End }
+func (r ScheduleDayOfWeekRange) GetStep() int32  { return r.Step }
+
+// ScheduleCalendarSpec is an event specification relative to the calendar, similar to a traditional cron specification.
+// A timestamp matches if at least one range of each field matches the
+// corresponding fields of the timestamp, except for year: if year is missing,
+// that means all years match. For all fields besides year, at least one Range must be present to match anything.
+type ScheduleCalendarSpec struct {
+	// Second range to match (0-59).
+	// Defaults to 0.
+	//
+	// +optional
+	// +kubebuilder:default={{"start": 0}}
+	Second []ScheduleSecondMinuteRange `json:"second,omitempty"`
```

Comment:
Originally I did indeed create a generic `ScheduleRange`. I changed to use this approach primarily as it allowed specifying `kubebuilder` validations where each range type only supports certain values. For example `ScheduleMonthRange` only allows the values 1 to 12 etc. I'm quite new to GO so if there is a cleaner way to do this i'm happy to review the solution. 

I also specified specific defaults for each range type as without defining certain fields on structured calendar temporal did not correctly create the schedule. I could move the the defaults to the controller instead rather than the CRD though I felt it was clearer to show the defaults on the CRD as it helps with troubleshooting.